### ISSUE TITLE
LogSumExp Optimisation

### ISSUE THREAD

**Issue by mgermain** [Theano/Theano#1563]
It could be nice to have a numerically more stable LogSumExp built-in.
Instead of doing T.log(T.sum(T.exp(x))) we could use something like LogSumExp(x).
```
def LogSumExp(x):
    x_max = T.max(x)
    T.log(T.sum(T.exp(x - x_max))) + x_max 
```
Or even better with axis support.
```
def LogSumExp(x, axis=None):
    x_max = T.max(x, axis=axis, keepdims=True)
    return T.log(T.sum(T.exp(x - x_max), axis=axis, keepdims=True)) + x_max
```

**Comment by nouiz** [comment]
This implementation is already in Pylearn: 

http://deeplearning.net/software/pylearn2/library/expr.html?highlight=log_sum_exp#pylearn2.expr.basic.log_sum_exp

There is a PR that add test for it: https://github.com/lisa-lab/pylearn2/pull/1390

Maybe we can just port it to Theano.

**Comment by mgermain** [comment]
I might be wrong but as I understand the scope of Theano, I think It would
make sense to have something basic like that in it instead of Pylearn.

**Comment by abergeron** [comment]
Yes, we want an optimization for that in theano.  This is the point of ticket #2082.

**Comment by nouiz** [comment]
Agreed, we just need someone to do it.
Le 16 févr. 2015 16:03, "Mathieu Germain" <notifications@github.com> a
écrit :

**Comment by mgermain** [comment]
"an optimization for that in theano"

Is there an example I could look at to learn how it should be implemented ?

**Comment by NeilGirdhar** [comment]
+1 This would be very useful.

**Comment by lamblin** [comment]
Closed by #4736
